frontend/src/utils/HistoryRecorder.js

```diff
@@ -1,55 +1,69 @@
-   * If there is only 1, it would be the root, and
-   * we can't go back any further.
+   * We need at least 1 items in the backwardHistory to go back
+   * @returns {bool} whether going back is valid
    */
   canGoBackward() {
-    return this.backwardHistory.length >= 2;
+    return this.backwardHistory.length >= 1;
   }
-
+  /**
+   * We need at least 1 items in the forwardHistory to go forward
+   * @returns {bool} whether going forward is valid
+   */
   canGoForward() {
-    return this.forwardHistory.length > 0;
+    return this.forwardHistory.length >= 1;
   }
 
-  resetPath(path) {
-    this.backwardHistory = [...path];
+  /**
+   * Adds a node to the backward history
+   * @param {Node} nodeName : Node to add to the backward history
+   */
+  addBackwardHistory(nodeName) {
+    if (
+      this.backwardHistory.length === 0 ||
+      this.backwardHistory[this.backwardHistory.length - 1] !== nodeName
+    ) {
+      this.backwardHistory.push(nodeName);
+    }
     this.forwardHistory = [];
-    this.onChange([...this.backwardHistory]);
   }
 
   /**
-   * If nodeName is passed as a parameter, go back up to that node.
+   * Pops and returns the previously visited node; adds the currently
+   * root node to the forward history so that the forward history works.
+   * @param {String} currentRootName: The name of the currently visible root
+   * @returns {String} previously visited node name
    */
-  goBackward(nodeName = null) {
-    if (!this.canGoBackward()) return;
-
-    do {
-      // Pop the last node from the current path and add
-      // it to the forwardHistory
-      this.forwardHistory.push(this.backwardHistory.pop());
-    } while (
-      // Go back more than once only if they passed this argument to the function
-      nodeName !== null &&
-      this.canGoBackward() &&
-      this.backwardHistory[this.backwardHistory.length - 1] !== nodeName
-    );
-
-    this.onChange([...this.backwardHistory]);
+  goBackward(currentRootName) {
+    if (!this.canGoBackward()) return "";
+    // Pop the last node from the current path and add
+    // it to the forwardHistory
+    const previouslyVisitedNodeName = this.backwardHistory.pop();
+    this.forwardHistory.push(currentRootName);
+    return previouslyVisitedNodeName;
   }
 
-  goForward() {
-    if (!this.canGoForward()) return;
-    this.backwardHistory.push(this.forwardHistory.pop());
-    this.onChange([...this.backwardHistory]);
+  /**
+   * Pops and returns the forward visited node; adds the currently
+   * root node to the backward history so that the backward history works.
+   * @param {String} currentRootName The name of the currently visible root
+   * @returns {String} the name of the next node in the forward history
+   */
+  goForward(currentRootName) {
+    if (!this.canGoForward()) return "";
+    if (
+      this.backwardHistory.length === 0 ||
+      this.backwardHistory[this.backwardHistory.length - 1] !== currentRootName
```

Comment:
Just curious, when would this be triggered?

Reply:
You're right, this case is prevented in the backward history function. Removed this condition